monix-tail/shared/src/main/scala/monix/tail/Iterant.scala

```diff
@@ -1888,6 +1888,38 @@ object Iterant extends IterantInstances {
   def pure[F[_], A](a: A): Iterant[F, A] =
     now[F, A](a)
 
+  /** Creates a stream that depends on resource allocated by a
+    * monadic value, ensuring the resource is released.
+    *
+    * Typical use-cases are working with files or network sockets
+    *
+    * @param acquire resource to acquire at the start of the stream
+    * @param use function that uses the resource to generate a stream of outputs
+    * @param release function that releases the acquired resource
+    *
+    * Example:
+    * {{{
+    *   val writeLines =
+    *     Iterant.bracket(IO { new PrintWriter("./lines.txt") })(
+    *       writer => Iterant[IO]
+    *         .fromIterator(Iterator.from(1))
+    *         .mapEval(i => IO { writer.println(s"Line #\$i") }),
+    *       (writer, _) => IO { writer.close() }
+    *     )
+    *
+    *   // Write 100 numbered lines to the file
+    *   // closing the writer when finished
+    *   writeLines.take(100).completeL.unsafeRunSync()
+    * }}}
+    */
+  def bracket[F[_], A, B](acquire: F[A])(
+    use: A => Iterant[F, B],
+    release: (A, BracketResult) => F[Unit]
```

Comment:
Btw, do you think it would make sense for `use` and `release` to be curried?
Maybe it's too much, but for `Task` it works with curly brackets, e.g..

```scala
task.bracket { in =>
  // use part
  ???
} { in =>
  // release part
  ???
}
```

Sort of resembles a `try/finally` statement to me and right now in Jan 2018 doesn't look ugly, but maybe I'll change my opinion in a month or so 😀

What do you think?

Reply:
I copied over the currying on `acquire` from fs2 because without it inference does not work.

Semantics of this `Iterant.bracket` are different from `try/finally` and bracket discussed at `cats-effect`, as here you can continue `flatMap`ping safely. Actually, because of this property, my `use` blocks in fs2 bracket are often one-liners:

```scala
fs2.Stream.bracket(delay { file.open() })(
  file => fs2.Stream(file),
  file => delay { file.close() }
)
```

I think it's nicer this way than

```scala
fs2.Stream.bracket(delay { file.open() }) { file =>
  fs2.Stream(file)
} { file =>
  delay { file.close() }
}
```

although I would not mind having it curried just for the sake of consistency, as `Task` has it this way and it seems to be where `cats-effect` is going too.

Reply:
I think you're right that the semantics are different and that the `use` function gets to be very light.

So leave it as it is.